README.md
# FowlerRecurringEvents
## Introduction
This library is an implementation of Martin Fowler's algorithms for recurring calendar events.
See https://martinfowler.com/apsupp/recurring.pdf for more information.

This library is divided into several "Temporal Expression" classes, each of which can be used to determine
whether an event occurs on a specific target date according to the specified pattern.

There are several different types of patterns included here which to the best of our knowledge are the most common
amongst mainstream calendar providers such as Google Calendar.

## Installation
To add this library into your project, run:
```
composer require moves/fowler-recurring-events
```

## Available Patterns
Note that all patterns also include a `frequency` option which allows you to specify how often the pattern applies.
For example, for the **Days of Week** pattern, you might want to specify recurrence every *other* weeks, rather than
every week. In this case, you would set a `frequency` of 2 to specify that the pattern applies every 2 weeks.

- **Day of Month** - Recurrence on a specific numbered day of every month
    - 1 = 1st day of every month
    - 10 = 10th day of every month
    - -1 = Last day of every month
    - -2 = 2nd to last day of every month
- **Day of Week of Month** - Recurrence on a specific weekday on a numbered week of every month
    - (3, 1) = 1st Wednesday of every month
    - (1, 2) = 2nd Monday of every month
    - (6, -1) = Last Saturday of every month
    - (7, -2) = 2nd to last Sunday of every month
- **Day of Year** - Recurrence on a specific calendar date every year
    - (25, 12) = December 25th of every year
    - (1, 4) = April 1st of every year
    - (29, 2) = February 29th on every leap year, March 1 on every non-leap year
- **Days** - Recurrence repeating on a certain number of days
    - 1 = Every day
    - 2 = Every other day
    - 10 = Every 10 days
- **Days of Week** - Recurrence on certain days of the week
    - [1, 2, 3, 4, 5] = Every weekday (Monday - Friday)
    - [6, 7] = Every weekend (Saturday - Sunday)
    - [1, 5] = Every Monday and Friday

## Usage
To determine whether your repeating event occurs on a particular target date, first instantiate an instance of the
appropriate Temporal Expression class with the details of your recurrence pattern, then call the `includes()` function.

Also note that all temporal expressions require a date on which the pattern begins.

For example, for an event which repeats every December 25th:
```
use Moves\FowlerRecurringEvents\TemporalExpressions\TEDayOfYear;

$patternStart = new DateTime('336-12-25'); //First recorded Christmas celebration
$eventPattern = new TEDayOfYear($patternStart, 25, 12);

$targetDate = new DateTime('2021-12-25');

$eventOccursOnTargetDate = $eventPattern->includes($targetDate);
// Expected Result: true

$otherTargetDate = new DateTime('2022-01-01');

$eventOccursOnOtherTargetDate = $eventPattern->includes($otherTargetDate);
// Expected Result: false
```

### Frequency and Pattern End Date
The frequency and pattern end date properties are optional when constructing a temporal expression.
To set values, for these optional properties, use the setter functions.
```
$eventPattern->setFrequency(2);
$eventPattern->setEndDate(new DateTime('2022-01-01'));
```

### Builder Pattern
You can use the builder pattern with setter method chaining for more convenience when building your temporal 
expressions.
```
use Moves\FowlerRecurringEvents\TemporalExpressions\TEDaysOfWeek;

$eventStart = new DateTime('2021-01-04');
$eventEnd = new DateTime('2021-12-27');
$eventPattern = TEDaysOfWeek::build($eventStart, 1)
  ->setEndDate($eventEnd)
  ->setFrequency(2);
```

## Practical Usage and Advice
The "Temporal Expression" algorithms are not intended to exist in a vacuum. In almost every instance, you will want
to store a single "master" instance for the recurrence pattern, usually the first occurrence, along with the type
of recurrence pattern or Temporal Expression and the required details of that particular pattern.

For developers who are inexperienced with calendar systems, it can be incredibly tempting to simply pre-calculate 
and store every possible instance of a repeating calendar event. However, doing so comes with massive inefficiencies
as the runtime *and* data storage complexity for creating, editing, and deleting a series of events grows linearly the 
further out you intend to project. Plus, using this method, your projection has to stop somewhere, otherwise you
would be stuck creating instances off into infinity, which in turn means that you must impose an end date to the pattern
on your users.

Alternatively, you might think improve the data storage complexity and circumvent the "end date" limitation by storing
a single "master" instance, and dynamically project that instance into the future. While this is the correct thought
process, it is another place where inexperienced developers often run astray. The temptation is often to simply
iterate over your pattern, day by day, week by week, month by month, or year by year, from the start date to the
target projection date. With this method, of course, the runtime grows linearly the further out you intend to project,
and thus it is not ideal for production calendar systems.

The Fowler methodology describes `O(1)` time algorithms for projecting calendar events, meaning that no matter how far
out you intend to project, the runtime is always the same. If you intend to project out a list of calendar events that 
will occur during a specified period of time, such as rendering a month at a time in a Calendar GUI, you can query a 
list of "master" events, then iterate over each day you want to create a projection for, checking the Temporal 
expression for each master event on each day to dynamically build your list.

There are, of course, many many optimizations that you can apply to reduce the number of cases you have to check and
calculate for when projecting a list of events for a specific period of time, but those are outside of the scope of
this library. 


CHANGELOG.md
<!--- BEGIN HEADER -->
# Changelog

All notable changes to this project will be documented in this file.
<!--- END HEADER -->

## [4.0.0-beta.10](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/compare/3.0.0...4.0.0-beta.10) (2023-08-01)

### ⚠ BREAKING CHANGES

* Change recurrence pattern end date to be exclusive, considering time ([59f5b1](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/commit/59f5b1fcff6a3471c5b201dacba8f723c02c837e))
* Store and respect timezone for all pattern dates ([28bb13](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/commit/28bb1347d347a82820a20c97037cd184f266101b))

### Bug Fixes

* Handle null end value in fromJson ([4214a4](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/commit/4214a4def53177ad3fe6d8791c935e5f182e45ff))
* Make new timezone field backwards-compatible ([423f3d](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/commit/423f3d5a76ff1e03210eea873f5f3b01e9491c9a), [009c53](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/commit/009c5382179059d256ba185d24ef6c5f5811fc70))
* Make return value for getEnd optional ([ba1af2](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/commit/ba1af22a4cc130b8078076ae81ce657e1145f889))
* Respect instance timezone when evaluating frequency from pattern start ([d738b1](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/commit/d738b18d46e50584136ac07cc6d957e1820d1db8))
* Respect timezone when evaluating includes on all patterns ([bbfe74](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/commit/bbfe7498de5a78a79179566546682bd7ed09eb06))


---

## [4.0.0-beta.9](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/compare/3.0.0...4.0.0-beta.9) (2023-07-27)

### ⚠ BREAKING CHANGES

* Change recurrence pattern end date to be exclusive, considering time ([59f5b1](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/commit/59f5b1fcff6a3471c5b201dacba8f723c02c837e))
* Store and respect timezone for all pattern dates ([28bb13](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/commit/28bb1347d347a82820a20c97037cd184f266101b))

### Bug Fixes

* Handle null end value in fromJson ([4214a4](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/commit/4214a4def53177ad3fe6d8791c935e5f182e45ff))
* Make new timezone field backwards-compatible ([423f3d](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/commit/423f3d5a76ff1e03210eea873f5f3b01e9491c9a), [009c53](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/commit/009c5382179059d256ba185d24ef6c5f5811fc70))
* Respect instance timezone when evaluating frequency from pattern start ([d738b1](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/commit/d738b18d46e50584136ac07cc6d957e1820d1db8))
* Respect timezone when evaluating includes on all patterns ([bbfe74](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/commit/bbfe7498de5a78a79179566546682bd7ed09eb06))


---

## [4.0.0-beta.8](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/compare/3.0.0...4.0.0-beta.8) (2023-07-26)

### ⚠ BREAKING CHANGES

* Change recurrence pattern end date to be exclusive, considering time ([59f5b1](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/commit/59f5b1fcff6a3471c5b201dacba8f723c02c837e))
* Store and respect timezone for all pattern dates ([28bb13](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/commit/28bb1347d347a82820a20c97037cd184f266101b))

### Bug Fixes

* Handle null end value in fromJson ([4214a4](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/commit/4214a4def53177ad3fe6d8791c935e5f182e45ff))
* Make new timezone field backwards-compatible ([423f3d](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/commit/423f3d5a76ff1e03210eea873f5f3b01e9491c9a), [009c53](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/commit/009c5382179059d256ba185d24ef6c5f5811fc70))
* Respect instance timezone when evaluating frequency from pattern start ([d738b1](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/commit/d738b18d46e50584136ac07cc6d957e1820d1db8))


---

## [4.0.0-beta.7](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/compare/3.0.0...4.0.0-beta.7) (2023-07-26)

### ⚠ BREAKING CHANGES

* Change recurrence pattern end date to be exclusive, considering time ([59f5b1](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/commit/59f5b1fcff6a3471c5b201dacba8f723c02c837e))
* Store and respect timezone for all pattern dates ([28bb13](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/commit/28bb1347d347a82820a20c97037cd184f266101b))

### Bug Fixes

* Handle null end value in fromJson ([4214a4](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/commit/4214a4def53177ad3fe6d8791c935e5f182e45ff))
* Make new timezone field backwards-compatible ([009c53](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/commit/009c5382179059d256ba185d24ef6c5f5811fc70))
* Respect instance timezone when evaluating frequency from pattern start ([d738b1](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/commit/d738b18d46e50584136ac07cc6d957e1820d1db8))


---

## [4.0.0-beta.6](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/compare/3.0.0...4.0.0-beta.6) (2023-07-25)

### ⚠ BREAKING CHANGES

* Change recurrence pattern end date to be exclusive, considering time ([59f5b1](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/commit/59f5b1fcff6a3471c5b201dacba8f723c02c837e))
* Store and respect timezone for all pattern dates ([28bb13](https://github.com/vetmoves/com.moves.php.fowler-recurring-events/commit/28bb1347d347a82820a20c97037cd184f266101b))

### Bug Fixes


tests/TestCases/TEDaysTest.php
<?php

namespace Tests\TestCases;

use Carbon\Carbon;
use Moves\FowlerRecurringEvents\TemporalExpressions\TEDays;
use PHPUnit\Framework\TestCase;

class TEDaysTest extends TestCase
{
    public function testCorrectDateBeforePatternStartReturnsFalse()
    {
        $pattern = TEDays::build(new Carbon('2021-01-01'));
        $testDate = new Carbon('2020-12-31');

        $result = $pattern->includes($testDate);

        $this->assertFalse($result);
    }

    public function testCorrectDateOnPatternStartReturnsTrue()
    {
        $pattern = TEDays::build(new Carbon('2021-01-01'));
        $testDate = new Carbon('2021-01-01');

        $result = $pattern->includes($testDate);

        $this->assertTrue($result);
    }

    public function testCorrectDateAfterPatternEndReturnsFalse()
    {
        $pattern = TEDays::build(new Carbon('2021-01-01'))
            ->setEndDate(new Carbon('2021-01-31'));
        $testDate = new Carbon('2020-02-01');

        $result = $pattern->includes($testDate);

        $this->assertFalse($result);
    }

    public function testCorrectDateOnPatternEndReturnsTrue()
    {
        $pattern = TEDays::build(new Carbon('2021-01-01'))
            ->setEndDate(new Carbon('2021-01-31'));
        $testDate = new Carbon('2021-01-31');

        $result = $pattern->includes($testDate);

        $this->assertTrue($result);
    }

    public function testBasicIncorrectDateReturnsFalse()
    {
        $pattern = TEDays::build(new Carbon('2021-01-01'))
            ->setFrequency(5);
        $testDate = new Carbon('2021-01-02');

        $result = $pattern->includes($testDate);

        $this->assertFalse($result);
    }

    public function testBasicCorrectDateReturnsTrue()
    {
        $pattern = TEDays::build(new Carbon('2021-01-01'));
        $testDate = new Carbon('2021-01-06');

        $result = $pattern->includes($testDate);

        $this->assertTrue($result);
    }
}


tests/TestCases/TEDayOfYearTest.php
<?php

namespace Tests\TestCases;

use Carbon\Carbon;
use Moves\FowlerRecurringEvents\TemporalExpressions\TEDayOfYear;
use PHPUnit\Framework\TestCase;

class TEDayOfYearTest extends TestCase
{
    public function testCorrectDateBeforePatternStartReturnsFalse()
    {
        $pattern = TEDayOfYear::build(new Carbon('2021-01-01'), 1, 1);
        $testDate = new Carbon('2020-01-01');

        $result = $pattern->includes($testDate);

        $this->assertFalse($result);
    }

    public function testCorrectDateOnPatternStartReturnsTrue()
    {
        $pattern = TEDayOfYear::build(new Carbon('2021-01-01'), 1, 1);
        $testDate = new Carbon('2021-01-01');

        $result = $pattern->includes($testDate);

        $this->assertTrue($result);
    }

    public function testCorrectDateAfterPatternEndReturnsFalse()
    {
        $pattern = TEDayOfYear::build(new Carbon('2020-01-01'), 1, 1)
            ->setEndDate(new Carbon('2021-01-01'));
        $testDate = new Carbon('2022-01-01');

        $result = $pattern->includes($testDate);

        $this->assertFalse($result);
    }

    public function testCorrectDateOnPatternEndReturnsTrue()
    {
        $pattern = TEDayOfYear::build(new Carbon('2020-01-01'), 1, 1)
            ->setEndDate(new Carbon('2021-01-01'));
        $testDate = new Carbon('2021-01-01');

        $result = $pattern->includes($testDate);

        $this->assertTrue($result);
    }

    public function testBasicIncorrectDateReturnsFalse()
    {
        $pattern = TEDayOfYear::build(new Carbon('2021-01-01'), 1, 1);
        $testDate = new Carbon('2021-12-25');

        $result = $pattern->includes($testDate);

        $this->assertFalse($result);
    }

    public function testBasicCorrectDateReturnsTrue()
    {
        $pattern = TEDayOfYear::build(new Carbon('2021-01-01'), 1, 1);
        $testDate = new Carbon('2022-01-01');

        $result = $pattern->includes($testDate);

        $this->assertTrue($result);
    }

    public function testCorrectDateWithIncorrectFrequencyReturnsFalse()
    {
        $pattern = TEDayOfYear::build(new Carbon('2021-01-01'), 1, 1)
            ->setFrequency(2);
        $testDate = new Carbon('2022-01-01');

        $result = $pattern->includes($testDate);

        $this->assertFalse($result);
    }

    public function testCorrectDateWithCorrectFrequencyReturnsTrue()
    {
        $pattern = TEDayOfYear::build(new Carbon('2021-01-01'), 1, 1)
            ->setFrequency(2);
        $testDate = new Carbon('2023-01-01');

        $result = $pattern->includes($testDate);

        $this->assertTrue($result);
    }

    public function testLeapDayMatchesMarch1OnLeapYearReturnsFalse()
    {
        $pattern = TEDayOfYear::build(new Carbon('2021-01-01'), 29, 2);
        $testDate = new Carbon('2024-03-01');

        $result = $pattern->includes($testDate);

        $this->assertFalse($result);
    }

    public function testLeapDayMatchesMarch1OffLeapYearReturnsTrue()
    {
        $pattern = TEDayOfYear::build(new Carbon('2021-01-01'), 29, 2);
        $testDate = new Carbon('2021-03-01');

        $result = $pattern->includes($testDate);

        $this->assertTrue($result);
    }
}


tests/TestCases/TEDayOfMonthTest.php
<?php

namespace Tests\TestCases;

use Carbon\Carbon;
use Moves\FowlerRecurringEvents\TemporalExpressions\TEDayOfMonth;
use PHPUnit\Framework\TestCase;

class TEDayOfMonthTest extends TestCase
{
    public function testCorrectDateBeforePatternStartReturnsFalse()
    {
        $pattern = TEDayOfMonth::build(new Carbon('2021-01-01'), 1);
        $testDate = new Carbon('2020-01-01');

        $result = $pattern->includes($testDate);

        $this->assertFalse($result);
    }

    public function testCorrectDateOnPatternStartReturnsTrue()
    {
        $pattern = TEDayOfMonth::build(new Carbon('2021-01-01'), 1);
        $testDate = new Carbon('2021-01-01');

        $result = $pattern->includes($testDate);

        $this->assertTrue($result);
    }

    public function testCorrectDateAfterPatternEndReturnsFalse()
    {
        $pattern = TEDayOfMonth::build(new Carbon('2020-01-01'), 1)
            ->setEndDate(new Carbon('2021-01-01'));
        $testDate = new Carbon('2021-02-01');

        $result = $pattern->includes($testDate);

        $this->assertFalse($result);
    }

    public function testCorrectDateOnPatternEndReturnsTrue()
    {
        $pattern = TEDayOfMonth::build(new Carbon('2020-01-01'), 1)
            ->setEndDate(new Carbon('2021-01-01'));
        $testDate = new Carbon('2021-01-01');

        $result = $pattern->includes($testDate);

        $this->assertTrue($result);
    }

    public function testBasicIncorrectDateReturnsFalse()
    {
        $pattern = TEDayOfMonth::build(new Carbon('2021-01-01'), 1);
        $testDate = new Carbon('2021-12-25');

        $result = $pattern->includes($testDate);

        $this->assertFalse($result);
    }

    public function testBasicCorrectDateReturnsTrue()
    {
        $pattern = TEDayOfMonth::build(new Carbon('2021-01-01'), 1);
        $testDate = new Carbon('2021-12-01');

        $result = $pattern->includes($testDate);

        $this->assertTrue($result);
    }

    public function testCorrectDateWithIncorrectFrequencyReturnsFalse()
    {
        $pattern = TEDayOfMonth::build(new Carbon('2021-01-01'), 1)
            ->setFrequency(2);
        $testDate = new Carbon('2021-12-01');

        $result = $pattern->includes($testDate);

        $this->assertFalse($result);
    }

    public function testCorrectDateWithCorrectFrequencyReturnsTrue()
    {
        $pattern = TEDayOfMonth::build(new Carbon('2021-01-01'), 1)
            ->setFrequency(2);
        $testDate = new Carbon('2021-11-01');

        $result = $pattern->includes($testDate);

        $this->assertTrue($result);
    }

    public function testCorrectDateFromEndWithIncorrectFrequencyReturnsFalse()
    {
        $pattern = TEDayOfMonth::build(new Carbon('2021-01-01'), -1)
            ->setFrequency(2);
        $testDate = new Carbon('2021-12-31');

        $result = $pattern->includes($testDate);

        $this->assertFalse($result);
    }

    public function testCorrectDateFromEndWithCorrectFrequencyReturnsTrue()
    {
        $pattern = TEDayOfMonth::build(new Carbon('2021-01-01'), -1)
            ->setFrequency(2);
        $testDate = new Carbon('2021-11-30');

        $result = $pattern->includes($testDate);

        $this->assertTrue($result);
    }

    public function testCorrectDateAcrossYearsWithIncorrectFrequencyReturnsFalse()
    {
        $pattern = TEDayOfMonth::build(new Carbon('2021-01-01'), 1)
            ->setFrequency(5);
        $testDate = new Carbon('2022-01-01'); //12 months later

        $result = $pattern->includes($testDate);

        $this->assertFalse($result);
    }

    public function testCorrectDateAcrossYearsWithCorrectFrequencyReturnsTrue()
    {
        $pattern = TEDayOfMonth::build(new Carbon('2021-01-01'), 1)
            ->setFrequency(5);
        $testDate = new Carbon('2022-04-01'); //15 months later

        $result = $pattern->includes($testDate);

        $this->assertTrue($result);
    }
}


src/TemporalExpressions/TEDays.php
<?php

namespace Moves\FowlerRecurringEvents\TemporalExpressions;

use Carbon\Carbon;
use DateTimeInterface;
use Moves\FowlerRecurringEvents\Contracts\ACTemporalExpression;

/**
 * Class TEDays
 * @package Moves\FowlerRecurringEvents\TemporalExpressions
 *
 * Temporal Expression for evaluating recurrence on a certain number of days.
 * E.g. "Every day", "Every other day", "Every 10 days"
 */
class TEDays extends ACTemporalExpression
{
    //region Setup
    /** @var string English representation of Temporal Expression type */
    public const TYPE = 'Days';

    /** @var int Number of days between repetitions */
    protected $frequency = 1;

    /**
     * TEDays constructor.
     * @param DateTimeInterface $start Starting date of repetition pattern
     */
    public function __construct(?DateTimeInterface $start)
    {
        parent::__construct($start);
    }

    /**
     * TEDays creator.
     * @param array $options
     * @return TEDays
     */
    public static function create(array $options): ACTemporalExpression
    {
        return static::build(
            isset($options['start']) 
                ? Carbon::create($options['start'])->setTimezone($options['timezone'] ?? 'UTC') 
                : null
        )->setupOptions($options);
    }

    /**
     * TEDays builder.
     * @param DateTimeInterface $start Starting date of repetition pattern
     * @return TEDays
     */
    public static function build(?DateTimeInterface $start): TEDays
    {
        return new static($start);
    }
    //endregion

    //region Iteration
    /**
     * @inheritDoc
     */
    public function next(): ?DateTimeInterface
    {
        if (is_null($this->current) || $this->current < $this->start)
        {
            $this->current = Carbon::create($this->start)->subDay();
        }

        $current = Carbon::create($this->current);
        $next = $current->copy()->addDay();

        while ((is_null($this->end) || $next < $this->end) && !$this->includes($next))
        {
            $daysToAdd = $this->frequency - ($next->diffInDays($this->start) % $this->frequency);
            $next->addDays($daysToAdd);
        }

        if (!is_null($this->end) && $next > $this->end) {
            $this->current = null;
        } else {
            $this->current = $next;
        }

        return $this->current;
    }
    //endregion

    //region Helpers
    protected function hasCorrectFrequencyFromStart(Carbon $instance): bool
    {
        $start = Carbon::create($this->start)
            ->setTime(0, 0);

        $instanceDay = Carbon::create($instance)
            ->setTimezone($start->timezone)
            ->setTime(0, 0);

        return $start->diffInDays($instanceDay) % $this->frequency == 0;
    }
    //endregion

    /**
     * @inheritDoc
     */
    public function includes(DateTimeInterface $date): bool
    {
        $start = (new Carbon($this->start));
        $end = is_null($this->end) ? null : (new Carbon($this->end));
        $instance = (new Carbon($date))->setTimezone($start->timezone);

        return $instance >= $start
            && (is_null($end) || $instance < $end)
            && $this->hasCorrectFrequencyFromStart($instance)
            && !$this->isIgnored($instance);
    }
}


src/TemporalExpressions/TEDayOfMonth.php
<?php

namespace Moves\FowlerRecurringEvents\TemporalExpressions;

use Carbon\Carbon;
use DateTimeInterface;
use Moves\FowlerRecurringEvents\Contracts\ACTemporalExpression;

/**
 * Class TEDayOfMonth
 * @package Moves\FowlerRecurringEvents\TemporalExpressions
 *
 * Temporal Expression for evaluating recurrence as a certain day of the month.
 * E.g. "1st of every month", "10th of every month", "Last day of every month"
 */
class TEDayOfMonth extends ACTemporalExpression
{
    //region Setup
    /** @var string English representation of Temporal Expression type */
    public const TYPE = 'Day of Month';

    /** @var int Day of month (positive from beginning of month, negative from end of month) */
    protected $dayOfMonth;

    /** @var int Number of months between repetitions */
    protected $frequency = 1;

    /**
     * TEDayOfMonth constructor.
     * @param DateTimeInterface $start Starting date of repetition pattern
     * @param int $dayOfMonth Day of month (positive from beginning of month, negative from end of month)
     */
    public function __construct(?DateTimeInterface $start, ?int $dayOfMonth)
    {
        parent::__construct($start);

        $this->dayOfMonth = $dayOfMonth;
    }

    /**
     * TEDayOfMonth creator.
     * @param array $options
     * @return TEDayOfMonth
     */
    public static function create(array $options): ACTemporalExpression
    {
        return static::build(
            isset($options['start']) 
                ? Carbon::create($options['start'])->setTimezone($options['timezone'] ?? 'UTC') 
                : null,
            $options['day_of_month'] ?? null
        )->setupOptions($options);
    }

    /**
     * TEDayOfMonth builder.
     * @param DateTimeInterface $start Starting date of repetition pattern
     * @param int $dayOfMonth Day of month (positive from beginning of month, negative from end of month)
     * @return TEDayOfMonth
     */
    public static function build(?DateTimeInterface $start, ?int $dayOfMonth): TEDayOfMonth
    {
        return new static($start, $dayOfMonth);
    }

    /**
     * @return array
     */
    public function toArray(): array
    {
        return array_merge(parent::toArray(), [
            'day_of_month' => $this->dayOfMonth,
        ]);
    }

    protected static function VALIDATION_RULES_TYPE(string $key = null): array
    {
        $prefix = empty($key) ? '' : "{$key}.";

        $class = static::TYPE;

        $requiredIfRule = "required_if:{$prefix}type,$class";

        return [
            $prefix . 'day_of_month' => "{$requiredIfRule}|integer|gte:-31|lte:31"
        ];
    }
    //endregion

    //region Getters
    public function getDayOfMonth(): int
    {
        return $this->dayOfMonth;
    }
    //endregion

    //region Iteration
    /**
     * @inheritDoc
     *
     * This iteration logic for this Temporal Expression type is necessarily incredibly complex,
     * and any future developers should avoid modifying the logic if possible.
     */
    public function next(): ?DateTimeInterface
    {
        /** Handle first iteration and manually set date before pattern start */
        if (is_null($this->current) || $this->current < $this->start)
        {
            /** Bring iteration date up to the day before the pattern start */
            $this->current = Carbon::create($this->start)->subDay();
        }

        /** Create Carbon instances, since they are easier to work with */
        $start = Carbon::create($this->start);
        $current = Carbon::create($this->current);
        $next = $current->copy();

        /** If the day of month is positive (i.e. from the beginning of the month) */
        if ($this->dayOfMonth > 0)
        {
            /**
             * If the pattern date has already passed for this month,
             * jump to the fist day of the next month.
             * Warning: Do not use "addMonth", because of the possibility
             * of skipping over shorter months.
             * e.g. 2021-01-31 with "addMonth" jumps to 2021-03-03
             */
            if ($next->day >= $this->dayOfMonth) {
                $next->setDay($next->daysInMonth);
                $next->addDay();
            }

            /** Iterate until the end of the pattern passes, or we find a valid date */
            while ((is_null($this->end) || $next < $this->end) && !$this->includes($next))
            {
                /** Move with the first day of the currently selected month */
                $next->setDay(1);

                /**
                 * From the current month, jump to the next month in the pattern
                 * Special case for the first date in the pattern: ignore this jump
                 */
                if ($next > $this->start) {
                    /**
                     * Jump to the next month which has a date in the pattern.
                     * Warning: Do not use "diffInMonths" because of the possibility
                     * of being off by 1, because it calculates based on the length of the month,
                     * rather than by the "month" component of the date.
                     * e.g. 2021-01-31 "diffInMonths" to 2021-03-15 is 1, not 2
                     */
                    $diffInMonths = (($next->year - $start->year) * 12) + $next->month - $start->month;
                    $monthsToAdd = $this->frequency - ($diffInMonths % $this->frequency);
                    $next->addMonths($monthsToAdd);
                }

                /**
                 * We should now be in the correct month, so
                 * jump to the next valid day in the pattern
                 */
                $next->addDays($this->dayOfMonth - 1);
            }
        }

        /** If the day of month is negative (i.e. from the end of the month) */
        else
        {
            /**
             * Determine the correct day for the current month.
             * e.g. If $this->dayOfMonth is -1, then for January 2021, the correct day is 31,
             * but for February 2021, the correct day is 28.
             * Note: $this->dayOfMonth is negative, so using the addition operator
             * is correct to achieve subtraction of the correct number of days from the
             * end of the current month.
             */
            $targetDay = $next->daysInMonth + 1 + $this->dayOfMonth;

            /**
             * If the pattern date has already passed for this month,
             * jump to the fist day of the next month.
             * Warning: Do not use "addMonth", because of the possibility
             * of skipping over shorter months.
             * e.g. 2021-01-31 with "addMonth" jumps to 2021-03-03
             */
            if ($next->day >= $targetDay) {
                $next->setDays($next->daysInMonth);
                $next->addDay();

                /** Also be sure to re-calculate the target day, since the month changed. */
                $targetDay = $next->daysInMonth + 1 + $this->dayOfMonth;
            }

            /** Jump to the target day for the current month */
            $next->setDay($targetDay);

            /** Iterate until the end of the pattern passes, or we find a valid date */
            while ((is_null($this->end) || $next < $this->end) && !$this->includes($next))
            {
                /** Move with the first day of the currently selected month */
                $next->setDay(1);

                /**
                 * From the current month, jump to the next month in the pattern
                 * Special case for the first date in the pattern: ignore this jump
                 */
                if ($next > $this->start)
                {
                    /**
                     * Jump to the next month which has a date in the pattern.
                     * Warning: Do not use "diffInMonths" because of the possibility
                     * of being off by 1, because it calculates based on the length of the month,
                     * rather than by the "month" component of the date.
                     * e.g. 2021-01-31 "diffInMonths" to 2021-03-15 is 1, not 2
                     */
                    $diffInMonths = (($next->year - $start->year) * 12) + $next->month - $start->month;
                    $monthsToAdd = $this->frequency - ($diffInMonths % $this->frequency);
                    $next->addMonths($monthsToAdd);
                }

                /**
                 * We should now be in the correct month, so
                 * jump to the next valid day in the pattern.
                 * Note: $this->dayOfMonth is negative, so using the addition operator
                 * is correct to achieve subtraction of the correct number of days from the
                 * end of the current month.
                 */
                $next->addDays($next->daysInMonth + $this->dayOfMonth);
            }
        }

        /**
         * Finally, we should have he next date in the pattern, but we have to check that we
         * have not run past the end date. If we have run past the end date, end the iteration
         * and return null. Otherwise, set the $this->current tracker and return the next valid date.
         */
        if (!is_null($this->end) && $next > $this->end) {
            $this->current = null;
        } else {
            $this->current = $next;
        }

        return $this->current;
    }
    //endregion

    //region Helpers


    protected function dayFromStartMatches(Carbon $instance): bool
    {
        return $this->dayOfMonth == $instance->day;
    }

    protected function dayFromEndMatches(Carbon $instance): bool
    {
        $daysInMonth = $instance->daysInMonth;
        return ($daysInMonth + 1) - abs($this->dayOfMonth) == $instance->day;
    }

    protected function hasCorrectFrequencyFromStart(Carbon $instance): bool
    {
        $start = Carbon::create($this->start)
            ->setTime(0, 0);

        $instanceDay = Carbon::create($instance)
            ->setTimezone($start->timezone)
            ->setTime(0, 0);

        $diffInYears = $instanceDay->year - $start->year;
        $diffInMonths = $instanceDay->month + (12 * $diffInYears) - $start->month;

        return $diffInMonths % $this->frequency == 0;
    }
    //endregion

    /**
     * @inheritDoc
     */
    public function includes(DateTimeInterface $date): bool
    {
        $start = (new Carbon($this->start));
        $end = is_null($this->end) ? null : (new Carbon($this->end));
        $instance = (new Carbon($date))->setTimezone($start->timezone);

        return $instance >= $start
            && (is_null($end) || $instance < $end)
            && (
                $this->dayOfMonth > 0 ?
                $this->dayFromStartMatches($instance) :
                $this->dayFromEndMatches($instance)
            )
            && $this->hasCorrectFrequencyFromStart($instance)
            && !$this->isIgnored($instance);
    }
}
